### PR TITLE
MotionPro

### PR DESCRIPTION
**Pull request by bilioicik**
test

### CODE REVIEW

**Comment by bilioicik** [comment]
Finally, I know how to pr.

**Comment by bilioicik** [comment]
Here are MotionPro and OperatorsPro. They have been done, but I can still improve them in some ways. And I plan to make more Pro after all.

**Comment by Arkos123** [comment]
image
image
我不李姐这些积木的意义是什么（）
这个下拉菜单有点让人费解，另外true/false个人认为应该做成菱形积木
image
image
这些积木让人有点费解（比如作为原型，本质这些表达）很怪（）
image

**Comment by MoreBugOfDog** [comment]
Oh, no! This is the same extension I already submitted for review! @bilioicik I already submitted this extension! Gotta close this PR or it will affect the reviewer's review.

**Comment by bilioicik** [comment]
@Arkos123 就是注入了私有变量“移动速度”“移动方向”“转动速度”，为了表述简洁，我就写了两个字，不过这些在文档里交代清楚了。这样移动转动，省去了传参过程，对运行总是好的。

**Comment by bilioicik** [comment]
@Arkos123 哦，你的“菱形积木”原来是这个意思。我以为你是说其他积木的圆形框。不过在条件的参数里塞定值本身就没有意义。这个积木主要作用是把字符串的“true”“false”这类转为真的布尔值的true、false，好push到数组对象。

**Comment by bilioicik** [comment]
@Arkos123 （我知道对象不能用push函数，前一句说混在一起了）反正就是帮忙更改数组对象里的元素为true、false、undefined这种特殊类型值。扩展文档里有专门提及这个积木的用途。

**Comment by bilioicik** [comment]
@MoreBugOfDog I know this, but I think pushing to my pr is more convernient because I still want to improve sth.

**Comment by bilioicik** [comment]
@MoreBugOfDog I know this, but I think pushing to my pr is more convenient because I still want to improve something in my extensions.

**Comment by bilioicik** [comment]
@MoreBugOfDog So that I don't need to disturb you again and again.

**Comment by bilioicik** [comment]
@sylarhcn 我新开了一个请求，关于改良过的运动Pro和运算Pro，望过审。

**Comment by MoreBugOfDog** [comment]
> 这样我就不需要一次又一次地打扰你了。
But you spelled my name wrong.
Correct:
多bug的啸天犬 @ CCW
Incorrect:
MoreBugOfDog @ CCW

**Comment by bilioicik** [comment]
> > 这样我就不需要一次又一次地打扰你了。
> > But you spelled my name wrong.
> > Correct:
> > 多bug的啸天犬 @ CCW
> > Incorrect:
> > MoreBugOfDog @ CCW

All right. I think English username is cooler. Actually, 凌 use FurryR in his extension.

**Comment by bilioicik** [comment]
> > But you spelled my name wrong.
> > Correct:
> > 多bug的啸天犬 @ CCW
> > Incorrect:
> > MoreBugOfDog @ CCW

All right. I think English username is cooler. Actually, 凌 used FurryR in his extension.

**Comment by bilioicik** [comment]
@MoreBugOfDog I will change it tomorrow.

**Comment by bilioicik** [comment]
@MoreBugOfDog So maybe you need to close your pr. Do you want to make more changes in OperatorsPro?

**Comment by fath11** [comment]
Please create your own folder and put the extensions in that folder
image

**Comment by fath11** [comment]
Doing this is terrible; I prefer the old version of this extension 
image

This will force the user to set the movement speed, etc., with another block before being able to move. In typical projects, the speed will stay the same, but with more advanced projects, such as a custom 3D renderer or projects that need to do a lot of heavy calculations, not being able to input values to blocks directly will significantly impact performance.

**Comment by bilioicik** [comment]
@fath11 The old blocks of this extension are still here. To avoid clutter, they are hidden. Press the second button then you can see them. 
These new blocks can be used in many different ways. If the types of their arguments are NUMBER, they will directly use the values which you put. You can also put variables'names or id there, they will find the variables you want.

**Comment by bilioicik** [comment]
@fath11 OK, I created my own folder just now.

**Comment by fath11** [comment]
movement speed, movement direction, and rotation angle are all just variables. There is no reason to make the block palate bigger. It is cool to have togglable blocks, but why try to hide the clutter when the clutter can be removed?

**Comment by bilioicik** [comment]
@fath11 The old blocks are more useful for their running speed because they do not recognize types of values. So I don’t want to remove them. But the number of them is plenty big. It’s bad for editing. Users won’t feel good. So l add them together to new blocks.

**Comment by bilioicik** [comment]
image
@fath11 Would you like this? I made blocks shorter and easy to understand.

**Comment by bilioicik** [comment]
@sylarhcn @fath11 请核审

**Comment by bilioicik** [comment]
@sylarhcn @fath11 请核审

**Comment by fath11** [comment]
> image @fath11 Would you like this? I made blocks shorter and easy to understand.

The use case for variables is really small. There are other kind of variables from other extensions. Implementing and maintaining support for every single kind of variables in every extensions is impossible. Sure you can just drop the reporter of that variable into any of the blocks and it works just fine, but then you might as well use normal blocks with normal input parameters.

You need to make a compromise between minor performance increase and usableness/not making the block palette uselessly large and **the extension's code uselessly complicated to maintain**.

**Comment by bilioicik** [comment]
It’s hard for me to choose one. Well, in my opinion, it’s ok to have both. I have made several tests to make sure there’s no bug in this extension. All of the blocks can work properly and are easy to use. On one hand, it is true that we can use Scratch blocks to achieve some similar aims. On the other hand,this will let very young and new Scratchers do it easily. Don’t we have a section called “Beginners”? It is reasonable to let beginners use it. Maybe some of them like using single menu blocks, others want to have better running speed. Perhaps it’s not necessary to create some blocks, but it’s more unnecessary to remove them. I think no one among ordinary users will refuse to have more choices.

By the way, I have expected for a long time. It will make me mad to go on to change it during Grade 9 such a busy term. Of course, it’s because of my poor experience.

**Comment by fath11** [comment]
How about make a button to switch between both kind of blocks instead of how it is right now. So only one kind of the block will appear at any one time and the rest are hidden

**Comment by bilioicik** [comment]
About the variable menu, if users have the same advice as yours, I will take it at once. Only the type of the argument will be changed. No problem with the value. So you don’t need to worry about the changing after it is public will cause bad influence. And you don’t need to pay attention to this before it is public.

**Comment by bilioicik** [comment]
I tried to make a button like that before, but it seems that there’s a bug in Gandi. “! variable”in blocks will cause ERROR.

**Comment by bilioicik** [comment]
I mean, the method like “hidefrompallette: ! this.hideextrablocks”.

**Comment by bilioicik** [comment]
“hidefrompallette: ! this.hideextrablocks” in blocks doesn’t work . Strangely, “hidefrompallette: ! this.hideextrablocks” in buttons always works.

**Comment by fath11** [comment]
split the blocks into two arrays, when the button is pressed, switch which array is being given to the blocks property and refresh the block palette

**Comment by bilioicik** [comment]
Ok, I have done it. But in other ways.

**Comment by bilioicik** [comment]
@sylarhcn Please review.

File: bilioicik/OperatorsPro
Comment (MoreBugOfDog) [inline bilioicik/OperatorsPro:1]: > link:   https://ccw.site/student/{{oid}}  =>  https://ccw.site/student/{{uid}}

这样会更好？
缩短了链接长度，但这不是必须的，我个人认为uid比乱码id更好看些
示例
>link:   https://www.ccw.site/student/6200811f05660557606c8b15  => https://www.ccw.site/student/197354885

**Review by MoreBugOfDog — COMMENTED** [review]
可以看一下我对链接美化的建议，这是我的个人喜好，非必须

File: bilioicik/ControlPro [hunk 1]
```diff
@@ -0,0 +1,1008 @@
...
+        }
+        rTest(args,util){
+            return 
+        }
+        cmTest(args,util){
+            util.target.stopAll()
+        }
+
+        isDefinedStrictly(a){
+            ! [0,false,undefined,null,'','0','false','undefined','null'].includes(a)
```
Comment (sylarhcn) [inline bilioicik/ControlPro:708]: 这种数组你应该定义为静态的常量来使用

**Review by sylarhcn — CHANGES_REQUESTED** [review]
1. 文件使用正确的后缀
2. 一个 pr 请只包含一个扩展的修改，
3. 过于复杂的扩展，请在 pr 详情中，写明扩展的设计目的和积木的功能，积木可以使用截图，这样能更好的被讨论与审核

**Comment by bilioicik** [comment]
改了后缀，正在研究怎么分开交，我前面新开了个ControlPro，但之前的内容意外出现在那儿的commits里了。

**Comment by sylarhcn** [comment]
> 改了后缀，正在研究怎么分开交，我前面新开了个ControlPro，但之前的内容意外出现在那儿的commits里了。

那我先关闭这个 pr。
你可以开多个开发分支，然后单独 pr 就可以。
最后你在合并的开发分支就行了。

**Comment by bilioicik** [comment]
image
21e6c4067fece829d5827bdb7416de12

用多个branch不太明白，我想应该是新创建一个branch，但情况咋跟这教程不太一样呢…

**Comment by bilioicik** [comment]
我打算先把其他文件删了保留一个

**Pull request closed without merging**